### PR TITLE
Fix: Update visualizations with multi-datasource support

### LINKED ISSUE #691: Add multi-data source aware visualizations

### PR DESCRIPTION
**Pull request by cythrawll**
Resolves #691 

## Description

make sure visualizations get the right namespace to pull namespaced stuff out of metadata services.

## Proposed Changes

- Add metaNamespace field to navi-viz-c3-chart
- Add namespace support to metric name service and metric name helpers
- Make sure all metadata calls in visualizations get a namespace passed in.
- chart-data/data utils can now group on more fields besides id.
- moved defaultDataSource in navi-app to prevent  error message.

## License

I confirm that this contribution is made under an MIT license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: packages/data/addon/helpers/metric-format.js [hunk 1]
```diff
@@ -25,15 +26,15 @@ export default Helper.extend({
...
+  compute([metric, namespace = getDefaultDataSourceName() /*...rest*/]) {
     let longName = '--';
     if (!metric) {
       return longName;
     }
 
     let metricId = get(metric, 'metric');
     if (isPresent(metricId)) {
-      longName = get(this, 'metricName').getLongName(metricId);
+      longName = get(this, 'metricName').getLongName(metricId, namespace);
```
Comment (kevinhinterlong) [inline packages/data/addon/helpers/metric-format.js]:
```suggestion
      longName = this.metricName.getLongName(metricId, namespace);
```

File: packages/data/addon/services/metric-name.js [hunk 2]
```diff
@@ -18,18 +18,18 @@ export default Service.extend({
    * @param {String} metricId - base metric name for a metric
    * @returns {String} - long name for the metric from the metadata
    */
-  getLongName(metricId) {
-    return get(this, 'metricMeta').getMetaField('metric', metricId, 'longName', metricId);
+  getLongName(metricId, namespace) {
+    return get(this, 'metricMeta').getMetaField('metric', metricId, 'longName', metricId, namespace);
```
Comment (kevinhinterlong) [inline packages/data/addon/services/metric-name.js]:
```suggestion
    return this.metricMeta.getMetaField('metric', metricId, 'longName', metricId, namespace);
```

File: packages/data/tests/integration/helpers/metric-format-test.js [hunk 3]
```diff
@@ -43,4 +43,21 @@ module('helper:metric-format', function(hooks) {
     this.set('metric', { metric: 'foo' });
     assert.dom().hasText('foo');
   });
+
+  test('multi-datasource support', async function(assert) {
```
Comment (kevinhinterlong) [inline packages/data/tests/integration/helpers/metric-format-test.js]: Add test for falling back to default

File: packages/core/addon/utils/data.js [hunk 4]
```diff
@@ -47,7 +47,36 @@ export function mostRecentData(rows) {
...
+}
+
+/**
+ * Selects best available dimension field to group on.
+ *
+ * @param {Array} rows
+ * @param {String} dimension
+ * @returns {String} field
+ */
+export function bestDimensionField(rows, dimension) {
```
[the remark this answers is not in the document]
Comment (cythrawll) [reply packages/core/addon/utils/data.js]: I think at the time I was writing this. it seemed more consistent to pass the whole rowset in as the rest of the other functions do in this utility.  I could be convinced to refactor though if you feel strongly about it.

I'm not sure about doing such a heavy assert/warn, also they won't be unique if say, the dataset is across several days.

File: packages/core/tests/unit/utils/data-test.js [hunk 5]
```diff
@@ -187,4 +187,48 @@ module('Unit | Utils | Data', function() {
...
+      bestDimensionField(
+        [
+          {
+            'dim|desc': 'foo'
+          }
+        ],
+        'dim'
+      ),
+      'dim|desc',
+      'desc is prefferred when it is the only field available'
```
[the remark this answers is not in the document]
Comment (cythrawll) [reply packages/core/tests/unit/utils/data-test.js]: F

File: packages/core/tests/integration/components/navi-visualizations/table-test.js [hunk 6]
```diff
@@ -183,6 +184,28 @@ module('Integration | Component | table', function(hooks) {
     );
   });
 
+  test('render alternative datasource', async function(assert) {
+    assert.expect(2);
+    const bardMeta = this.owner.lookup('service:bard-metadata');
+    bardMeta._keg.reset();
+    await bardMeta.loadMetadata({ dataSourceName: 'blockhead' });
+    const blockheadModel = cloneDeep(Model[0]);
+    blockheadModel.request.dataSource = 'blockhead';
```
[the remark this answers is not in the document]
Comment (cythrawll) [reply packages/core/tests/integration/components/navi-visualizations/table-test.js]: The metadata asserts that the metadata for that datasource hasn't been loaded and the table does not render.  Sure I can add a test.

[the remark this answers is not in the document]
Comment (cythrawll) [reply packages/core/tests/integration/components/navi-visualizations/table-test.js]: actually ember expectAssertion doesn't support async renders yet (pr open for 6 mo now). And did a quick trial of trying to get it working without it and ended up yak shaving so we'll need to come back to this one.

File: packages/core/addon/components/navi-visualizations/pie-chart.js [hunk 7]
```diff
@@ -98,11 +100,11 @@ class NaviVisualizationsPieChartComponent extends Component.extend(hasChartBuild
   /**
    * @property {String} metricDisplayName - display name for metric
    */
-  @computed('options')
+  @computed('options', 'namespace')
   get metricDisplayName() {
     let metric = get(this, 'seriesConfig.metric');
 
-    return metric && get(this, 'metricName').getDisplayName(metric);
+    return metric && get(this, 'metricName').getDisplayName(metric, this.namespace);
```
Comment (jkusa) [inline packages/core/addon/components/navi-visualizations/pie-chart.js]:
```suggestion
    return metric && this.metricName.getDisplayName(metric, this.namespace);
```

File: packages/core/addon/components/navi-visualizations/table.js [hunk 8]
```diff
@@ -177,14 +177,16 @@ class Table extends Component {
...
    * @returns {Boolean}
    */
-  _hasCustomDisplayName(column) {
+  _hasCustomDisplayName(column, namespace) {
     if (isBlank(column.displayName)) {
       return false;
     }
 
-    let defaultName = getColumnDefaultName(column, get(this, 'bardMetadata'));
+    let defaultName = getColumnDefaultName(column, get(this, 'bardMetadata'), namespace);
```
Comment (jkusa) [inline packages/core/addon/components/navi-visualizations/table.js]:
```suggestion
    const defaultName = getColumnDefaultName(column, this.bardMetadata, namespace);
```

File: packages/core/addon/helpers/default-column-name.js [hunk 9]
```diff
@@ -47,7 +49,7 @@ export function getColumnDefaultName({ type, attributes }, bardMetadata) {
 export default Helper.extend({
   bardMetadata: service(),
 
-  compute([column]) {
-    return getColumnDefaultName(column, get(this, 'bardMetadata'));
+  compute([column, namespace]) {
+    return getColumnDefaultName(column, get(this, 'bardMetadata'), namespace);
```
Comment (jkusa) [inline packages/core/addon/helpers/default-column-name.js]:
```suggestion
    return getColumnDefaultName(column, this.bardMetadata, namespace);
```

File: packages/core/addon/utils/chart-data.js [hunk 10]
```diff
@@ -121,7 +124,7 @@ export function buildDimensionSeriesValues(request, rows) {
       let id = get(row, `${dimension}|id`),
         desc = get(row, `${dimension}|desc`);
 
-      values[dimension] = id;
+      values[dimension] = id || desc;
```
Comment (jkusa) [inline packages/core/addon/utils/chart-data.js]: Do we have a test for this?

File: packages/core/addon/utils/data.js [hunk 11]
```diff
@@ -47,7 +47,39 @@ export function mostRecentData(rows) {
...
+    return fields[0];
+  }
+  //TODO: add meta dataservice to get dimension meta data id field, need to get namespace down here consistently though
+  const preferredFields = ['key', 'id', 'desc'];
+  for (const field of preferredFields) {
+    if (fields.includes(`${dimension}|${field}`)) {
+      return `${dimension}|${field}`;
+    }
+  }
+  return `${dimension}|id`;
```
Comment (jkusa) [inline packages/core/addon/utils/data.js]: By this time didn't we already check `${dimension}|id`? Should this be`fields[0]` which would some field of the dimension.

File: packages/core/tests/integration/components/navi-visualizations/goal-gauge-test.js [hunk 12]
```diff
@@ -45,6 +45,29 @@ module('Integration | Component | goal gauge ', function(hooks) {
...
+    const metaData = this.owner.lookup('service:bard-metadata');
+    metaData._keg.reset();
+    await metaData.loadMetadata({ dataSourceName: 'blockhead' });
+
+    _setModel(this, 'pageViews', 3030000000, 'blockhead');
+    set(this, 'metric', { metric: 'pageViews', paramters: {} });
+    await render(hbs`
+    <NaviVisualizations::GoalGauge
+       @model={{this.model}}
+        @options={{hash
```
Comment (jkusa) [inline packages/core/tests/integration/components/navi-visualizations/goal-gauge-test.js]:
```suggestion
       @options={{hash
```

File: packages/core/addon/utils/chart-data.js (hunk 10 above)
Comment (jkusa) [reply packages/core/addon/utils/chart-data.js]: Should we use the `bestDimensionField` method?

File: packages/core/addon/utils/data.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (jkusa) [reply packages/core/addon/utils/data.js]: What about:
```suggestion
export function getDimensionGroupingField(rows, dimension) {
```

File: packages/core/addon/utils/data.js (hunk 11 above)
Comment (cythrawll) [reply packages/core/addon/utils/data.js]: good point

**Pull request merged**